### PR TITLE
Session Timeout Task

### PR DESCRIPTION
**Pull request by velimd**
### JIRA link ###
https://tools.hmcts.net/jira/browse/NFDIV-482
https://tools.hmcts.net/jira/browse/NFDIV-359

### Change description ###
- call '/active' endpoint when user is active to add to extend session maxage
- if user is inactive for 20mins, call post to save data and then redirect to timed-out endpoint

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/app/controller/GetController.test.ts
Comment (linusnorton) [inline src/main/app/controller/GetController.test.ts:22]: shouldn't this default to false?

File: src/main/assets/js/session-timeout.ts
Comment (linusnorton) [inline src/main/assets/js/session-timeout.ts:20]: maybe it would be better to make this async and use await?

File: src/main/app/controller/GetController.test.ts
Comment (velimd) [reply src/main/app/controller/GetController.test.ts:22]: default should be true as you want to see the sign in/out button but in this specific page, its not there hence defaut is true.

if youre talking about in the actual get class, it does compare to see if the url is '/timed-out'

File: src/main/assets/js/session-timeout.ts
Comment (velimd) [reply src/main/assets/js/session-timeout.ts:20]: just for my understanding, why would that be better? i dont have a preference really but just thought the .then block makes it easier to read and as its combined together.

Comment (linusnorton) [reply src/main/assets/js/session-timeout.ts:20]: No technical difference, just personal preference I guess. I prefer to stick to the `const someVariable = something()` look.

For example:

```
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-type': 'application/json',
      'csrf-token': body._csrf as string,
    },
    body: JSON.stringify(body),
  });

  if (response.redirected) {
    window.location.href = response.url;
  }
```

**Review by linusnorton — APPROVED** [review]

File: src/main/steps/common/page.njk
Comment (SebFlippence) [inline src/main/steps/common/page.njk:20]: Rather than scattering `hideNavigationButton` throughout the code, it might be better to leave this as it is and change the block in the affected template.njk so that these links don't get rendered. For example: [#281/src/main/steps/save-sign-out/template.njk](https://github.com/hmcts/nfdiv-frontend/pull/281/files#diff-2ccf0cd072b3a6562cd536fa373ceed4eefc8dafb09ae346ceba79024d6e0786R3-R13)

**Review by SebFlippence — APPROVED** [review]

File: src/main/assets/js/session-timeout.ts
Comment (SebFlippence) [inline src/main/assets/js/session-timeout.ts:1]: For webpack tree shaking you might want to change this to:

```suggestion
import { throttle } from 'lodash';
```

This way webpack should only bundle `throttle` to the frontend rather than including the whole of lodash.

File: src/main/app/controller/PostController.ts
Comment (SebFlippence) [inline src/main/app/controller/PostController.ts:35]: Perhaps this branch could be tested in `PostController.test`?

File: src/main/assets/js/session-timeout.ts
Comment (SebFlippence) [inline src/main/assets/js/session-timeout.ts:42]: Would be nice to get some unit tests on this. I think you should be able to trigger a `keypress`, `click`, `mousemove` etc using Jest and then using Jest mock timers to advance the clock by 20 minutes you should be able to check if `fetch` was called with the right arguments.

Comment (velimd) [reply src/main/assets/js/session-timeout.ts:42]: this is becoming a bit of a pain hence i might merge this once you approve and do this in another pr

**Review by SebFlippence — APPROVED** [review]

**Pull request merged**
